File: python/src/acedatacloud/resources/grok.py
"""Grok resources — chat completions and video generation."""

from __future__ import annotations

import json as _json
from collections.abc import Iterator
from typing import Any, Literal

from .._runtime.tasks import AsyncTaskHandle, TaskHandle

GrokChatModel = Literal["grok-4.5", "grok-4", "grok-3"]
GrokVideoModel = Literal[
    "grok-imagine-video-1.5-fast:reverse",
    "grok-imagine-video:reverse",
    "grok-imagine-video:official",
    "grok-imagine-video-1.5:official",
    "grok-imagine-video",
]


def _task_id(result: Any) -> str:
    if not isinstance(result, dict):
        return ""
    if result.get("task_id"):
        return str(result["task_id"])
    data = result.get("data")
    if isinstance(data, dict) and data.get("task_id"):
        return str(data["task_id"])
    return str(result.get("id") or "")


class _GrokCompletions:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    def create(
        self,
        *,
        model: GrokChatModel | str,
        messages: list[dict[str, Any]],
        stream: bool = False,
        **kwargs: Any,
    ) -> dict[str, Any] | Iterator[dict[str, Any]]:
        body = {"model": model, "messages": messages, **kwargs}
        if stream:
            body["stream"] = True
            return self._stream(body)
        return self._transport.request("POST", "/grok/chat/completions", json=body)

    def _stream(self, body: dict[str, Any]) -> Iterator[dict[str, Any]]:
        for chunk in self._transport.request_stream("POST", "/grok/chat/completions", json=body):
            yield _json.loads(chunk)


class _AsyncGrokCompletions:
    def __init__(self, transport: Any) -> None:
        self._transport = transport

    async def create(
        self,
        *,
        model: GrokChatModel | str,
        messages: list[dict[str, Any]],
        stream: bool = False,
        **kwargs: Any,
    ) -> dict[str, Any]:
        body = {"model": model, "messages": messages, **kwargs}
        if stream:
            body["stream"] = True
            return self._stream(body)
        return await self._transport.request("POST", "/grok/chat/completions", json=body)

    async def _stream(self, body: dict[str, Any]):
        async for chunk in self._transport.request_stream("POST", "/grok/chat/completions", json=body):
            yield _json.loads(chunk)


class _GrokChatNamespace:
    def __init__(self, transport: Any) -> None:
        self.completions = _GrokCompletions(transport)


class _AsyncGrokChatNamespace:
    def __init__(self, transport: Any) -> None:
        self.completions = _AsyncGrokCompletions(transport)


class Grok:
    """Synchronous Grok client — chat completions and video generation."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport
        self.chat = _GrokChatNamespace(transport)

    def generate(
        self,
        *,
        prompt: str | None = None,
        model: GrokVideoModel | None = None,
        image_url: str | None = None,
        reference_image_urls: list[str] | None = None,
        aspect_ratio: Literal["1:1", "16:9", "9:16", "4:3", "3:4", "3:2", "2:3"] | None = None,
        resolution: Literal["480p", "720p", "1080p"] | None = None,
        duration: int | None = None,
        callback_url: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> TaskHandle:
        """Grok video generation."""
        body: dict[str, Any] = {**kwargs}
        if prompt is not None:
            body["prompt"] = prompt
        if model is not None:
            body["model"] = model
        if image_url is not None:
            body["image_url"] = image_url
        if reference_image_urls is not None:
            body["reference_image_urls"] = reference_image_urls
        if aspect_ratio is not None:
            body["aspect_ratio"] = aspect_ratio
        if resolution is not None:
            body["resolution"] = resolution
        if duration is not None:
            body["duration"] = duration
        if callback_url is not None:
            body["callback_url"] = callback_url
        if async_ is not None:
            body["async"] = async_
        result = self._transport.request("POST", "/grok/videos", json=body)
        return TaskHandle(_task_id(result), "/grok/tasks", self._transport, submitted=result)


class AsyncGrok:
    """Async Grok client — chat completions and video generation."""

    def __init__(self, transport: Any) -> None:
        self._transport = transport
        self.chat = _AsyncGrokChatNamespace(transport)

    async def generate(
        self,
        *,
        prompt: str | None = None,
        model: GrokVideoModel | None = None,
        image_url: str | None = None,
        reference_image_urls: list[str] | None = None,
        aspect_ratio: Literal["1:1", "16:9", "9:16", "4:3", "3:4", "3:2", "2:3"] | None = None,
        resolution: Literal["480p", "720p", "1080p"] | None = None,
        duration: int | None = None,
        callback_url: str | None = None,
        async_: bool | None = None,
        **kwargs: Any,
    ) -> AsyncTaskHandle:
        """Grok video generation."""
        body: dict[str, Any] = {**kwargs}
        if prompt is not None:
            body["prompt"] = prompt
        if model is not None:
            body["model"] = model
        if image_url is not None:
            body["image_url"] = image_url
        if reference_image_urls is not None:
            body["reference_image_urls"] = reference_image_urls
        if aspect_ratio is not None:
            body["aspect_ratio"] = aspect_ratio
        if resolution is not None:
            body["resolution"] = resolution
        if duration is not None:
            body["duration"] = duration
        if callback_url is not None:
            body["callback_url"] = callback_url
        if async_ is not None:
            body["async"] = async_
        result = await self._transport.request("POST", "/grok/videos", json=body)
        return AsyncTaskHandle(_task_id(result), "/grok/tasks", self._transport, submitted=result)

File: typescript/src/resources/grok.ts
/** Grok resources — chat completions and video generation. */

import { Transport } from '../runtime/transport';
import { TaskHandle } from '../runtime/tasks';

export const GROK_CHAT_MODELS = ['grok-4.5', 'grok-4', 'grok-3'] as const;
export type GrokChatModel = (typeof GROK_CHAT_MODELS)[number];

export const GROK_VIDEO_MODELS = [
  'grok-imagine-video-1.5-fast:reverse',
  'grok-imagine-video:reverse',
  'grok-imagine-video:official',
  'grok-imagine-video-1.5:official',
  'grok-imagine-video',
] as const;
export type GrokVideoModel = (typeof GROK_VIDEO_MODELS)[number];

function taskId(result: Record<string, unknown>): string {
  if (typeof result?.task_id === 'string') return result.task_id;
  const data = result?.data as Record<string, unknown> | undefined;
  if (data && typeof data.task_id === 'string') return data.task_id;
  return typeof result?.id === 'string' ? result.id : '';
}

class GrokCompletions {
  constructor(private transport: Transport) {}

  async create(opts: {
    model: GrokChatModel | string;
    messages: Array<Record<string, unknown>>;
    stream?: false;
    [key: string]: unknown;
  }): Promise<Record<string, unknown>>;
  async create(opts: {
    model: GrokChatModel | string;
    messages: Array<Record<string, unknown>>;
    stream: true;
    [key: string]: unknown;
  }): Promise<AsyncGenerator<Record<string, unknown>>>;
  async create(opts: {
    model: GrokChatModel | string;
    messages: Array<Record<string, unknown>>;
    stream?: boolean;
    [key: string]: unknown;
  }): Promise<Record<string, unknown> | AsyncGenerator<Record<string, unknown>>> {
    const { model, messages, stream, ...rest } = opts;
    const body: Record<string, unknown> = { model, messages, ...rest };
    if (stream) {
      body.stream = true;
      return this.streamResponse(body);
    }
    return this.transport.request('POST', '/grok/chat/completions', { json: body });
  }

  private async *streamResponse(body: Record<string, unknown>): AsyncGenerator<Record<string, unknown>> {
    for await (const chunk of this.transport.requestStream('POST', '/grok/chat/completions', { json: body })) {
      yield JSON.parse(chunk);
    }
  }
}

class GrokChat {
  readonly completions: GrokCompletions;
  constructor(transport: Transport) {
    this.completions = new GrokCompletions(transport);
  }
}

export class Grok {
  readonly chat: GrokChat;

  constructor(private transport: Transport) {
    this.chat = new GrokChat(transport);
  }

  async generate(opts: {
    prompt?: string;
    model?: GrokVideoModel | string;
    imageUrl?: string;
    referenceImageUrls?: string[];
    aspectRatio?: '1:1' | '16:9' | '9:16' | '4:3' | '3:4' | '3:2' | '2:3';
    resolution?: '480p' | '720p' | '1080p';
    duration?: number;
    callbackUrl?: string;
    async?: boolean;
    [key: string]: unknown;
  }): Promise<TaskHandle> {
    const { prompt, model, imageUrl, referenceImageUrls, aspectRatio, resolution, duration, callbackUrl, ...rest } = opts;
    const body: Record<string, unknown> = { ...rest };
    if (prompt !== undefined) body.prompt = prompt;
    if (model !== undefined) body.model = model;
    if (imageUrl !== undefined) body.image_url = imageUrl;
    if (referenceImageUrls !== undefined) body.reference_image_urls = referenceImageUrls;
    if (aspectRatio !== undefined) body.aspect_ratio = aspectRatio;
    if (resolution !== undefined) body.resolution = resolution;
    if (duration !== undefined) body.duration = duration;
    if (callbackUrl !== undefined) body.callback_url = callbackUrl;
    const result = (await this.transport.request('POST', '/grok/videos', { json: body })) as Record<string, unknown>;
    return new TaskHandle(taskId(result), '/grok/tasks', this.transport, result);
  }
}
